toolchain/check/import_ref.cpp

```diff
@@ -1643,8 +1652,14 @@ class ImportRefResolver {
 
     SemIR::ImplId impl_id = SemIR::ImplId::Invalid;
     if (!impl_const_id.is_valid()) {
-      // For impl loading, we can always start at the second phase. Create a
-      // forward declaration of the impl for any recursive references.
+      if (HasNewWork()) {
+        // This is the end of the first phase. Don't make a new impl yet if we
+        // already have new work.
+        return Retry();
+      }
```

Comment:
When is there new work here? I thought work could only be added by TryResolveTypedInst, and I'm not seeing what in the code above would result in new work. For example, comparing to other similar entity logic, this doesn't resolve a specific before checking HasNewWork.

Reply:
In principle, new work can be added by `TryResolveInst` before we even get here. (This would happen when resolving a generic constant that refers to an impl, which is actually not possible because we never form a generic constant that refers to an impl, but this code shouldn't be assuming that.)

Reply:
Maybe this should `CARBON_CHECK(!HasNewWork())` instead?

Reply:
It could be, but I think that couples this code unnecessarily to the assumption that we don't do any extra work before we get here, so I'd prefer this as-is.